File: dsa/tree/postorderTraversal.js
/**
 * Definition for a binary tree node.
 */
export class TreeNode {
	/**
	 * @param {number} value 
	 * @description : Initializes a binary tree node with a value, and left and right children set to null.
	 */
	constructor(value) {
		this.value = value;
		this.left = null;
		this.right = null;
	}
}

/**
 * 
 * @param {TreeNode} root : The root of the binary tree (and it's subtrees in recursive calls)
 * @returns {void} : Prints the value of each node in postorder sequence
 * @description : In postorder traversal, we visit the left subtree, then the right subtree, and finally the root node.
 */
export function postorderTraversal(root) {
	if (!root) return;
	postorderTraversal(root.left);
	postorderTraversal(root.right);
	console.log(root.value);
}

/*
Example Tree:

        4
      /   \  
     /     \
    2       5
   / \       \
  1   3       6
*/

// tiny self-checks in console
if (typeof window !== "undefined")
(function selfTest() {
const root = new TreeNode(4);
    root.left = new TreeNode(2);
    root.left.left = new TreeNode(1);
    root.left.right = new TreeNode(3);
    root.right = new TreeNode(5);
    root.right.right = new TreeNode(6);
    console.log("Pre-order Traversal of the tree:");
    postorderTraversal(root); // 1 3 2 6 5 4
})();

File: dsa_in_java/tree/PostorderTraversal.java
/*
 * Topic: Binary Tree Postorder Traversal
 * Language: Java
 *
 * Description:
 * Traverse a binary tree in postorder fashion: Left → Right → Root.
 *
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 * LeetCode Link: https://leetcode.com/problems/binary-tree-postorder-traversal/
 */

import java.util.*;

public class PostorderTraversal {

    public static void postorder(TreeNode root, List<Integer> result) {
        if (root == null) return;
        postorder(root.left, result);
        postorder(root.right, result);
        result.add(root.val);
    }

    public static List<Integer> postorderTraversal(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        postorder(root, result);
        return result;
    }

    public static void main(String[] args) {
        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(3);

        System.out.println("Postorder Traversal: " + postorderTraversal(root));
    }
}

/*
Example Output:
Postorder Traversal: [2, 3, 1]
*/
